### PR TITLE
Add TVHeadend widget

### PR DESCRIPTION
**Pull request by elParaguayo**
Not sure how much of a demand there is for this but I thought I'd share as I like it!

It's a small icon widget that shows when your TVHeadend server is recording and also lists the upcoming recordings in a popup window.

tvh_widget

### CODE REVIEW

File: libqtile/widget/tvheadend.py [hunk 1]
```diff
@@ -0,0 +1,327 @@
...
+
+class TVHWidget(base._Widget, base.MarginMixin):
+    """
+    A widget to show whether a TVHeadend server is currently recording or not.
+
+    The widget will also show a popup displaying upcoming recordings.
+
+    NB if you use a username and password, these are stored in plain text. You
+    may therefore wish to create an unprivileged user account in TVHeadend that
+    only has access to scheduled recordings data.
```
Comment (tych0) [inline libqtile/widget/tvheadend.py:99]: Can you add a note here that this depends on the third party module requests? We've had complaints about this before...

File: libqtile/widget/tvheadend.py [hunk 2]
```diff
@@ -0,0 +1,327 @@
...
+    """Get the path to tv icon"""
+    dir_path = Path(__file__).resolve() / ".." / ".." / "resources" / "tvheadend-icons"
+    return str(dir_path.resolve())
+
+
+class TVHTimeOut(Exception):
+    pass
+
+
+class TVHJobServer(object):
```
Comment (tych0) [inline libqtile/widget/tvheadend.py:46]: No need to inherit from object here.

File: libqtile/widget/tvheadend.py [hunk 3]
```diff
@@ -0,0 +1,327 @@
...
+
+import requests
+from requests.auth import HTTPBasicAuth
+
+from libqtile import bar, images
+from libqtile.log_utils import logger
+from libqtile.popup import Popup
+from libqtile.widget import base
+
+UPCOMING_RECORDINGS = "/dvr/entry/grid_upcoming"
```
Comment (tych0) [inline libqtile/widget/tvheadend.py:33]: Any chance this needs to be configurable?

File: libqtile/widget/tvheadend.py [hunk 4]
```diff
@@ -0,0 +1,327 @@
...
+    def _send_api_request(self, path, args=None):
+        url = self.host + path
+        try:
+            r = requests.post(url,
+                              data=args,
+                              auth=self.auth,
+                              timeout=self.timeout)
+            return r.json(strict=False)
+        except (requests.exceptions.Timeout, requests.ConnectionError):
+            raise TVHTimeOut()
```
Comment (tych0) [inline libqtile/widget/tvheadend.py:62]: This re-raising swallows the cause for the original exception, can we just drop this?

File: libqtile/widget/tvheadend.py [hunk 5]
```diff
@@ -0,0 +1,327 @@
...
+
+        if not self.data:
+            lines.append("No upcoming recordings.")
+
+        else:
+            lines.append("Upcoming recordings:")
+            for rec in self.data:
+                lines.append(self.popup_format.format(**rec))
+
+        self.popup = Popup(self.qtile,
```
Comment (tych0) [inline libqtile/widget/tvheadend.py:298]: By my count this is the first in-tree use of the popup code, nice :)

**Comment by tych0** [comment]
Overall looks good to me, modulo a few small things. Any chance there's a nice way to write one simple test? :)

**Comment by elParaguayo** [comment]
Wow. Quick reviews @tych0 !

All perfectly valid points so let me update for them later.

As for the popup, it's fantastic. My Strava widget uses it too but that one's waaaay too messy to upload!

File: libqtile/widget/tvheadend.py (hunk 2 above)
Comment (elParaguayo) [reply libqtile/widget/tvheadend.py:46]: Old habits die hard...

**Comment by tych0** [comment]
You made a strava widget?! Sounds awesome :)

File: libqtile/widget/tvheadend.py (hunk 3 above)
Comment (elParaguayo) [reply libqtile/widget/tvheadend.py:33]: _Shouldn't _be__ as it's behind the scenes and I don't think users can amend the path but I guess TVH could change their API so I can shift it to being configurable.

**Comment by elParaguayo** [comment]
Yes, you can see the Strava icon (or my attempt at drawing it) in the screenshot next to the TV. Clicking on it shows my running stats. It can also display text in the bar. It only works for running but it could probably be made to work for cycling too.

**Comment by elParaguayo** [comment]
As for writing a test, won't this fail as `requests` is not a required library?

**Comment by tych0** [comment]
You can put it in requirements-dev.txt; everyone should have that when running the tests.

**Comment by elParaguayo** [comment]
The test will only make sense if I can get the popup to work in the test. I've tried running the `test_popup.py` file locally with `pytest` but I get errors so I can't adapt it for my needs here. I get this message:
`ERROR test/test_popup.py::test_focus[BareConfig] - FileNotFoundError: [Errno 2] No such file or directory: 'Xvfb'`

Any ideas what I need to do to get that to work for me?

**Comment by tych0** [comment]
Install Xvfb :-)

**Comment by elParaguayo** [comment]
Ah. Assumed all packages would have been installed as part of dev requirements. Never heard of xvfb before. Learning lots as I go along...

**Comment by elParaguayo** [comment]
Sorry. I must have missed a step. It looks like test failed because it couldn't import requests but I did include it in the `requirements-dev.txt` file.

**Comment by tych0** [comment]
Huh, I thought tox would have installed that. Maybe try it with,

```
diff --git a/tox.ini b/tox.ini
index 16fd6616..3fed157d 100644
--- a/tox.ini
+++ b/tox.ini
@@ -25,6 +25,7 @@ deps =
     setuptools >= 40.5.0
     xcffib >= 0.8.1
 commands =
+    pip install -r requirements-dev.txt
     python3 setup.py pytest --addopts "-Wall --cov libqtile --cov-report term-missing {posargs}"
 
 [testenv:packaging]
```

**Comment by elParaguayo** [comment]
Thanks. That fixed it. Tests running successfully now.

**Comment by tych0** [comment]
lgtm, although there's a small conflict; can you rebase and squash in the review changes?

**Comment by elParaguayo** [comment]
Don't _think_ that check error is to do with my code.

**Comment by tych0** [comment]
Yeah, agreed. We've seen this failure on master just today. Something changed recently and is causing a lot of issues...

**Comment by elParaguayo** [comment]
Is there any way to re-run the tests?

**Comment by tych0** [comment]
Yes, although the failure rate looks to be ~50%, so without a lot of babysitting I'm not sure it'll do much good.

**Comment by elParaguayo** [comment]
Ah, ok. That's a pretty bad rate!

**Comment by tych0** [comment]
Yeah, I'll probably wait another week or two to debug. Hopefully the cairo people figure it out by then :)

**Comment by elParaguayo** [comment]
Added escaping of popup text to prevent pango parse_error (as per #1685)

**Comment by tych0** [comment]
Can you rebase this? I think CI is fixed and the code itself lgtm.

**Comment by elParaguayo** [comment]
Looks like my test has broken. Has something changed in the way tests are run?

**Comment by elParaguayo** [comment]
My test seems to pass so I'm assuming something else is going wrong with the tests again.

**Comment by elParaguayo** [comment]
Might be worth holding off merging this until #2072 is merged.

Currently, I have to create a Popup that's big and then shrink it to the necessary size as the Drawer isn't resized. 2072 should fix that.

**Comment by elParaguayo** [comment]
The old version of this PR inlcuded `requests` in `requirements-dev.txt` as that was needed to run the test. It looks like `tox.ini` has changed slightly since then. Am I right in saying I should now just add `requests` to the `deps` section under `[testenv]`?

**Pull request closed without merging**